File: Sources/FotufilmImaging/PrintFrameRenderer.swift
#if canImport(CoreGraphics)
import Foundation
import CoreGraphics
import CoreText
#if canImport(FotufilmCore)
import FotufilmCore
#endif

/// One finishing pass for CPU/Metal developments, canvas, thumbnails and export.
/// Physical geometry stays fixed; a different crop is fitted into the aperture without resampling.
public enum PrintFrameRenderer {
    public struct Layout {
        public let size: CGSize
        /// Core Graphics coordinates, with the origin at the lower left.
        public let imageRect: CGRect
        public let pixelsPerMM: CGFloat
        public let rotated: Bool
    }

    public static func layout(width: Int, height: Int,
                              configuration: PrintFrameConfiguration) -> Layout {
        guard configuration.frame != .none else {
            return Layout(size: CGSize(width: width, height: height),
                          imageRect: CGRect(x: 0, y: 0, width: width, height: height),
                          pixelsPerMM: 1, rotated: false)
        }
        if configuration.frame == .emulsion {
            // A crop-following mount, not a claim of physical film or paper dimensions.
            let short = CGFloat(min(width, height))
            let horizontal = ceil(short * 0.095), vertical = ceil(short * 0.135)
            return Layout(size: CGSize(width: CGFloat(width) + 2 * horizontal,
                                       height: CGFloat(height) + 2 * vertical),
                          imageRect: CGRect(x: horizontal, y: vertical,
                                            width: CGFloat(width), height: CGFloat(height)),
                          pixelsPerMM: 1, rotated: false)
        }
        let material = materialGeometry(configuration)
        let aperture = material.aperture
        let rotated = aperture.width != aperture.height
            && (width > height) != (aperture.width > aperture.height)
        let scale = max(CGFloat(rotated ? height : width) / aperture.width,
                        CGFloat(rotated ? width : height) / aperture.height)
        let rawSize = CGSize(width: material.size.width * scale, height: material.size.height * scale)
        let size = CGSize(width: ceil(rotated ? rawSize.height : rawSize.width),
                          height: ceil(rotated ? rawSize.width : rawSize.height))
        let centre = rotated
            ? CGPoint(x: rawSize.height - aperture.midY * scale, y: aperture.midX * scale)
            : CGPoint(x: aperture.midX * scale, y: aperture.midY * scale)
        return Layout(size: size,
                      imageRect: CGRect(x: (centre.x - CGFloat(width) / 2).rounded(),
                                        y: (centre.y - CGFloat(height) / 2).rounded(),
                                        width: CGFloat(width), height: CGFloat(height)),
                      pixelsPerMM: scale, rotated: rotated)
    }

    public static func render(_ image: CGImage, configuration: PrintFrameConfiguration) -> CGImage? {
        guard configuration.frame != .none else { return image }
        let placement = layout(width: image.width, height: image.height, configuration: configuration)
        guard let space = image.colorSpace, space.model == .rgb,
              let context = CGContext(data: nil, width: Int(placement.size.width),
                                      height: Int(placement.size.height), bitsPerComponent: 16,
                                      bytesPerRow: 0, space: space,
                                      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
                                        | CGBitmapInfo.byteOrder16Little.rawValue)
        else { return nil }
        let material = materialGeometry(configuration)
        context.setFillColor(baseColor(configuration.baseRGB))
        context.fill(CGRect(origin: .zero, size: placement.size))
        context.saveGState()
        context.scaleBy(x: placement.pixelsPerMM, y: placement.pixelsPerMM)
        if placement.rotated {
            context.translateBy(x: material.size.height, y: 0)
            context.rotate(by: .pi / 2)
        }
        if configuration.frame == .emulsion {
            guard EmulsionBorderRenderer.draw(in: context, around: placement.imageRect) else { return nil }
        } else if configuration.frame == .paper {
            let scale = placement.pixelsPerMM
            let photo = placement.rotated
                ? CGRect(x: placement.imageRect.minY / scale,
                         y: material.size.height - placement.imageRect.maxX / scale,
                         width: placement.imageRect.height / scale, height: placement.imageRect.width / scale)
                : placement.imageRect.applying(CGAffineTransform(scaleX: 1 / scale, y: 1 / scale))
            drawLustre(in: context, size: material.size, excluding: photo)
        } else if let geometry = configuration.geometry {
            if let printing = configuration.edgePrinting {
                drawEdgePrinting(in: context, geometry: geometry, printing: printing,
                                 color: baseColor(configuration.edgeRGB))
            }
            drawPerforations(in: context, geometry: geometry)
            if geometry.isSheet, let code = configuration.sheetNotches {
                drawNotches(in: context, size: material.size, code: code)
            }
        }
        context.restoreGState()
        // Copy at integer pixel coordinates after the border is drawn. The original profile,
        // 16-bit depth and every photograph pixel survive, including P3 and HLG delivery.
        context.interpolationQuality = .none
        context.setBlendMode(.copy)
        context.draw(image, in: placement.imageRect)
        return context.makeImage()
    }

    private static func materialGeometry(_ configuration: PrintFrameConfiguration)
        -> (size: CGSize, aperture: CGRect) {
        if let g = configuration.geometry {
            return (CGSize(width: g.widthMM, height: g.heightMM),
                    CGRect(x: g.apertureX, y: g.apertureY,
                           width: g.apertureWidth, height: g.apertureHeight))
        }
        // A real 4 × 6 inch sheet cut from a paper roll, with a chosen 3 mm easel margin.
        // Sheet/crop size is a presentation choice; it is not an intrinsic size of the emulsion.
        return (CGSize(width: 152.4, height: 101.6),
                CGRect(x: 3, y: 3, width: 146.4, height: 95.6))
    }

    private static func baseColor(_ rgb: SIMD3<Float>) -> CGColor {
        CGColor(colorSpace: CGColorSpace(name: CGColorSpace.extendedLinearDisplayP3)!,
                components: [CGFloat(rgb.x), CGFloat(rgb.y), CGFloat(rgb.z), 1])!
    }

    /// The neutral scan-bed visible through the physical holes.
    private static var cutoutColor: CGColor { baseColor(SIMD3(repeating: 0.96)) }

    private static func drawEdgePrinting(in context: CGContext, geometry g: FilmBorderGeometry,
                                         printing: FilmEdgePrinting, color: CGColor) {
        context.saveGState()
        // In the reference strip x is transport and y is across. Turn that strip with its
        // perforations when the camera transports vertically; portrait rotation is outside.
        if !g.horizontalTransport {
            context.translateBy(x: g.widthMM, y: 0)
            context.rotate(by: .pi / 2)
        }
        let along = g.horizontalTransport ? g.widthMM : g.heightMM
        let across = g.horizontalTransport ? g.heightMM : g.widthMM
        context.clip(to: CGRect(x: 0, y: 0, width: along, height: across))
        // System vector lettering approximates the edge printer. No manufacturer font or
        // scanned asset is shipped. Paths avoid display-sized hinting at small physical sizes.
        let font = CTFontCreateWithName("Helvetica" as CFString, 100, nil)
        context.setFillColor(color)
        for mark in printing.marks {
            let attributes = [kCTFontAttributeName: font] as CFDictionary
            let string = CFAttributedStringCreate(nil, mark.text as CFString, attributes)!
            let line = CTLineCreateWithAttributedString(string)
            let path = CGMutablePath()
            for run in CTLineGetGlyphRuns(line) as! [CTRun] {
                let count = CTRunGetGlyphCount(run)
                var glyphs = [CGGlyph](repeating: 0, count: count)
                var positions = [CGPoint](repeating: .zero, count: count)
                CTRunGetGlyphs(run, CFRange(location: 0, length: 0), &glyphs)
                CTRunGetPositions(run, CFRange(location: 0, length: 0), &positions)
                let runFont = (CTRunGetAttributes(run) as NSDictionary)[kCTFontAttributeName] as! CTFont
                for i in 0..<count {
                    if let glyph = CTFontCreatePathForGlyph(runFont, glyphs[i], nil) {
                        path.addPath(glyph, transform: CGAffineTransform(translationX: positions[i].x,
                                                                        y: positions[i].y))
                    }
                }
            }
            let bounds = path.boundingBoxOfPath
            guard !bounds.isEmpty, !bounds.isNull else { continue }
            context.saveGState()
            context.translateBy(x: mark.xMM, y: mark.yMM)
            context.scaleBy(x: mark.widthMM / bounds.width, y: mark.heightMM / bounds.height)
            context.translateBy(x: -bounds.minX, y: -bounds.minY)
            context.addPath(path)
            context.fillPath()
            context.restoreGState()
        }
        context.restoreGState()
    }

    private static func drawPerforations(in context: CGContext, geometry g: FilmBorderGeometry) {
        guard let type = g.perforation else { return }
        let width: CGFloat, height: CGFloat, radius: CGFloat, edge: CGFloat
        switch type {
        case .kodakStandard: (width, height, radius, edge) = (2.794, 1.981, 0.51, 2.01)
        case .bellHowell: (width, height, radius, edge) = (2.794, 1.854, 0, 2.01)
        case .sixteen: (width, height, radius, edge) = (1.829, 1.270, 0.25, 0.914)
        case .superEight: (width, height, radius, edge) = (0.914, 1.143, 0.13, 0.51)
        }
        context.saveGState()
        // Put both still and motion perforations in coordinates across/along film transport.
        if g.horizontalTransport {
            context.translateBy(x: 0, y: g.heightMM)
            context.rotate(by: -.pi / 2)
        }
        let across = CGFloat(g.horizontalTransport ? g.heightMM : g.widthMM)
        let along = CGFloat(g.horizontalTransport ? g.widthMM : g.heightMM)
        context.clip(to: CGRect(x: 0, y: 0, width: across, height: along))
        let pitch = CGFloat(g.pitchMM)
        // 16 mm holes align with the frame line; Super 8 holes align with the frame centre.
        let first: CGFloat = type == .sixteen ? 0 : pitch / 2
        let rows: [CGFloat] = g.rows == 2 ? [edge, across - edge - width] : [edge]
        context.setFillColor(cutoutColor)
        for x in rows {
            var centre = first
            while centre <= along {
                let rect = CGRect(x: x, y: centre - height / 2, width: width, height: height)
                if type == .bellHowell {
                    // BH is a circle clipped by two parallel flats, not a rounded KS rectangle.
                    context.saveGState()
                    context.clip(to: rect)
                    context.fillEllipse(in: CGRect(x: x, y: centre - width / 2,
                                                   width: width, height: width))
                    context.restoreGState()
                } else {
                    context.addPath(CGPath(roundedRect: rect, cornerWidth: radius,
                                           cornerHeight: radius, transform: nil))
                    context.fillPath()
                }
                centre += pitch
            }
        }
        context.restoreGState()
    }

    private static func drawNotches(in context: CGContext, size: CGSize, code: SheetFilmNotchCode) {
        // The code occupies the upper right edge when the sheet is upright, emulsion facing us.
        // Manufacturer diagrams specify shapes/order, not absolute dimensions: a 20 mm code
        // span and 10 mm corner setback are presentation conventions, documented as such.
        let span: CGFloat = 20
        let start = size.width - 10 - span
        context.setFillColor(cutoutColor)
        for notch in code.notches {
            let rect = CGRect(x: start + CGFloat(notch.position) * span,
                              y: size.height - CGFloat(notch.depth) * span,
                              width: CGFloat(notch.width) * span,
                              height: CGFloat(notch.depth) * span * 2)
            context.fillEllipse(in: rect)
        }
    }

    private static func drawLustre(in context: CGContext, size: CGSize, excluding photo: CGRect) {
        // Fine resin-coated stipple, never cotton fibres or deckled paper. The visible finish
        // is a restrained procedural approximation, not measured surface microtopography.
        var random = MaterialRandom()
        context.setFillColor(CGColor(gray: 0.12, alpha: 0.035))
        let pitch: CGFloat = 0.17
        var y: CGFloat = 0
        while y < size.height {
            var x: CGFloat = 0
            while x < size.width {
                let point = CGPoint(x: x + random.next() * pitch, y: y + random.next() * pitch)
                if !photo.contains(point) {
                    context.fillEllipse(in: CGRect(x: point.x, y: point.y, width: 0.065, height: 0.065))
                }
                x += pitch
            }
            y += pitch
        }
    }

    private struct MaterialRandom {
        var state: UInt64 = 0x5052494E54534854
        mutating func next() -> CGFloat {
            state = state &* 6364136223846793005 &+ 1442695040888963407
            return CGFloat(state >> 40) / CGFloat(1 << 24)
        }
    }
}
#endif

File: web/src/print-frame-renderer.js
import { frameNoise, emulsionTexture, emulsionRim } from "./frame-texture.js";
import { yieldToBrowser } from "./yield.js";
const color = (rgb) => `rgb(${rgb.map((v) => v * 255).join(" ")})`;
const fill = (ctx, rect) =>
  ctx.fillRect(rect.x, rect.y, rect.width, rect.height);
const inset = (r, d) => ({
  x: r.x + d,
  y: r.y + d,
  width: r.width - 2 * d,
  height: r.height - 2 * d,
});
const ellipse = (ctx, x, y, width, height) => {
  ctx.beginPath();
  ctx.ellipse(
    x + width / 2,
    y + height / 2,
    width / 2,
    height / 2,
    0,
    0,
    2 * Math.PI,
  );
  ctx.fill();
};

function perforations(ctx, g, dimensions) {
  if (!g.perforation || !dimensions) return;
  const { width, height, radius, edge } = dimensions;
  ctx.save();
  if (g.horizontalTransport) {
    ctx.translate(0, g.heightMM);
    ctx.rotate(-Math.PI / 2);
  }
  const across = g.horizontalTransport ? g.heightMM : g.widthMM,
    along = g.horizontalTransport ? g.widthMM : g.heightMM;
  ctx.beginPath();
  ctx.rect(0, 0, across, along);
  ctx.clip();
  for (const x of g.rows === 2 ? [edge, across - edge - width] : [edge]) {
    for (
      let centre = g.perforation === "sixteen" ? 0 : g.pitchMM / 2;
      centre <= along;
      centre += g.pitchMM
    ) {
      if (g.perforation === "bellHowell") {
        ctx.save();
        ctx.beginPath();
        ctx.rect(x, centre - height / 2, width, height);
        ctx.clip();
        ellipse(ctx, x, centre - width / 2, width, width);
        ctx.restore();
      } else {
        ctx.beginPath();
        ctx.roundRect(x, centre - height / 2, width, height, radius);
        ctx.fill();
      }
    }
  }
  ctx.restore();
}
function edgePrinting(ctx, g, printing) {
  if (!printing) return;
  ctx.save();
  if (!g.horizontalTransport) {
    ctx.translate(g.widthMM, 0);
    ctx.rotate(Math.PI / 2);
  }
  ctx.beginPath();
  ctx.rect(
    0,
    0,
    g.horizontalTransport ? g.widthMM : g.heightMM,
    g.horizontalTransport ? g.heightMM : g.widthMM,
  );
  ctx.clip();
  // Like the native renderer, use system vector lettering; no scanned logos or invented roll codes.
  ctx.font = "100px Helvetica, Arial, sans-serif";
  ctx.textBaseline = "alphabetic";
  for (const mark of printing.marks) {
    const m = ctx.measureText(mark.text);
    const width = m.actualBoundingBoxLeft + m.actualBoundingBoxRight;
    const height = m.actualBoundingBoxAscent + m.actualBoundingBoxDescent;
    if (!width || !height) continue;
    ctx.save();
    ctx.translate(mark.xMM, mark.yMM + mark.heightMM);
    ctx.scale(mark.widthMM / width, -mark.heightMM / height);
    ctx.fillText(mark.text, m.actualBoundingBoxLeft, m.actualBoundingBoxAscent);
    ctx.restore();
  }
  ctx.restore();
}
function carrier(ctx, photo, width) {
  const outer = inset(photo, -width);
  const corners = [
    [outer.x, outer.y],
    [outer.x + outer.width, outer.y],
    [outer.x + outer.width, outer.y + outer.height],
    [outer.x, outer.y + outer.height],
  ];
  const normals = [
    [0, -1],
    [1, 0],
    [0, 1],
    [-1, 0],
  ];
  ctx.beginPath();
  for (let side = 0; side < 4; side++) {
    const from = corners[side],
      to = corners[(side + 1) % 4],
      length = Math.hypot(to[0] - from[0], to[1] - from[1]);
    const count = Math.max(2, Math.ceil(length / 0.5));
    for (let i = 0; i <= count; i++) {
      const t = i / count,
        along = length * t;
      const wobble =
        frameNoise(along * 0.35, side * 7.3, 409) +
        0.4 * frameNoise(along * 1.6, side * 3.1, 613);
      const amount = Math.max(
        -width * 0.25,
        Math.min(width * 0.25, wobble * 0.22),
      );
      const x = from[0] + (to[0] - from[0]) * t + normals[side][0] * amount;
      const y = from[1] + (to[1] - from[1]) * t + normals[side][1] * amount;
      if (side === 0 && i === 0) ctx.moveTo(x, y);
      else ctx.lineTo(x, y);
    }
  }
  ctx.closePath();
  ctx.fill();
}
async function lustre(ctx, size, photo, stale) {
  let state = 0x5052494e54534854n;
  const random = () => {
    state = BigInt.asUintN(
      64,
      state * 6364136223846793005n + 1442695040888963407n,
    );
    return Number(state >> 40n) / 16777216;
  };
  ctx.fillStyle = "rgb(31 31 31 / 0.035)";
  const pitch = 0.17;
  for (let row = 0, y = 0; y < size.height; row++, y += pitch) {
    if (row % 64 === 0) {
      await yieldToBrowser();
      if (stale()) return;
    }
    for (let x = 0; x < size.width; x += pitch) {
      const px = x + random() * pitch,
        py = y + random() * pitch;
      if (
        px < photo.x ||
        px > photo.x + photo.width ||
        py < photo.y ||
        py > photo.y + photo.height
      )
        ellipse(ctx, px, py, 0.065, 0.065);
    }
  }
}

/** One finishing path for previews, comparisons and exports. Photograph pixels are copied
 * at integer coordinates; only the intentional emulsion rim may cover their edges. */
export async function renderPrintFrame(image, plan, stale = () => false, omitPhoto = false) {
  if (!plan || plan.configuration.frame === "none") return image;
  const c = plan.configuration,
    p = plan.placement,
    m = plan.materialSize,
    palette = plan.palette;
  const canvas = document.createElement("canvas");
  canvas.width = p.size.width;
  canvas.height = p.size.height;
  const ctx = canvas.getContext("2d");
  if (!ctx)
    throw new Error(
      "The framed image is too large. Choose a smaller export size.",
    );
  ctx.fillStyle = color(c.slideMount ? palette.card : palette.base);
  ctx.fillRect(0, 0, canvas.width, canvas.height);
  // Shared native geometry is lower-left. Physical material drawing stays in that space.
  ctx.save();
  ctx.translate(0, canvas.height);
  ctx.scale(p.scale, -p.scale);
  if (p.rotated) {
    ctx.translate(m.height, 0);
    ctx.rotate(Math.PI / 2);
  }
  const r = p.image,
    s = p.scale;
  const photo = p.rotated
    ? {
        x: r.y / s,
        y: m.height - (r.x + r.width) / s,
        width: r.height / s,
        height: r.width / s,
      }
    : { x: r.x / s, y: r.y / s, width: r.width / s, height: r.height / s };
  if (c.slideMount) {
    const mount = c.slideMount,
      x = (mount.mountMM - mount.apertureWidth) / 2,
      y = (mount.mountMM - mount.apertureHeight) / 2;
    ctx.save();
    ctx.beginPath();
    ctx.roundRect(
      x,
      y,
      mount.apertureWidth,
      mount.apertureHeight,
      mount.cornerRadiusMM,
    );
    ctx.clip();
    ctx.fillStyle = color(palette.base);
    ctx.fillRect(x, y, mount.apertureWidth, mount.apertureHeight);
    ctx.fillStyle = "rgb(0 0 0 / .18)";
    ctx.fillRect(x, y + mount.apertureHeight - 0.35, mount.apertureWidth, 0.35);
    ctx.fillRect(x, y, 0.35, mount.apertureHeight);
    ctx.restore();
  } else if (c.sheet) {
    ctx.fillStyle = color(palette.rebate);
    if (c.sheet.rebateMM) carrier(ctx, photo, c.sheet.rebateMM);
    if (c.hasLustre)
      await lustre(ctx, m, inset(photo, -c.sheet.rebateMM), stale);
  } else if (c.geometry) {
    ctx.fillStyle = color(palette.edge);
    edgePrinting(ctx, c.geometry, c.edgePrinting);
    ctx.fillStyle = color(palette.cutout);
    perforations(ctx, c.geometry, plan.perforation);
    for (const notch of c.sheetNotches?.notches || [])
      ellipse(
        ctx,
        m.width - 30 + notch.position * 20,
        m.height - notch.depth * 20,
        notch.width * 20,
        notch.depth * 40,
      );
  }
  ctx.restore();
  if (stale()) return null;
  ctx.imageSmoothingEnabled = false;
  const top = canvas.height - r.y - r.height;
  if (omitPhoto) ctx.clearRect(r.x, top, r.width, r.height);
  else {
    // Integer placement preserves the delivered photograph without resampling.
    const pixels = image.getContext("2d").getImageData(0, 0, image.width, image.height);
    ctx.putImageData(pixels, r.x, top);
  }
  if (c.frame === "emulsion") {
    const texture = await emulsionTexture(r.width, r.height, stale);
    if (!texture) return null;
    const layer = document.createElement("canvas");
    layer.width = texture.width;
    layer.height = texture.height;
    layer
      .getContext("2d")
      .putImageData(
        new ImageData(texture.bytes, texture.width, texture.height),
        0,
        0,
      );
    const keep = Math.ceil(Math.min(r.width, r.height) * emulsionRim);
    ctx.save();
    ctx.beginPath();
    ctx.rect(0, 0, canvas.width, canvas.height);
    ctx.rect(
      r.x + keep,
      top + keep,
      Math.max(0, r.width - 2 * keep),
      Math.max(0, r.height - 2 * keep),
    );
    ctx.clip("evenodd");
    ctx.imageSmoothingEnabled = true;
    ctx.imageSmoothingQuality = "high";
    ctx.drawImage(
      layer,
      r.x - texture.fringe * texture.unit,
      top - texture.fringe * texture.unit,
      texture.width * texture.unit * texture.step,
      texture.height * texture.unit * texture.step,
    );
    ctx.restore();
  }
  return canvas;
}
